carla/recourse_methods/catalog/face/library/face_method.py

```diff
@@ -0,0 +1,282 @@
+            data,
+            immutable_constraint_matrix1,
+            immutable_constraint_matrix2,
+            index,
+            n,
+            y_positive_indeces,
+            is_knn=is_knn,
+        )
+
+    candidate_counterfactual_star = np.array(candidate_counterfactuals_star)
+
+    # STEP 4 -- COMPUTE DISTANCES between x^F and candidate x^CF; else return NaN
+    if candidate_counterfactual_star.size == 0:
+        candidate_counterfactual_star = np.empty(
+            data.values.shape[1],
+        )
+        candidate_counterfactual_star[:] = np.nan
+
+    else:
+        if p_norm == 1:
+            c_dist = np.abs((data.values[index] - candidate_counterfactual_star)).sum(
+                axis=1
+            )
+        elif p_norm == 2:
+            c_dist = np.square(
+                (data.values[index] - candidate_counterfactuals_star)
+            ).sum(axis=1)
+        else:
+            raise ValueError("Distance not defined yet. Choose p_norm to be 1 or 2")
+
+        min_index = np.argmin(c_dist)
+        candidate_counterfactual_star = candidate_counterfactual_star[min_index]
+
+    return candidate_counterfactual_star
+
+
+def choose_random_subset(data, frac, index):
+    """
+    Choose a subset of data for computational efficiency
+
+    Parameters
+    ----------
+    data : pd.DataFrame
+    frac: float 0 < number =< 1
+        fraction of data for which we compute the graph; if frac = 1, and data set large, then compute long
+    index: int
+
+    Returns
+    -------
+    pd.DataFrame
+    """
+    number_samples = np.int(np.rint(frac * data.values.shape[0]))
+    chosen_indeces = np.random.choice(
+        np.arange(0, index).tolist()
+        + np.arange(index + 1, data.values.shape[0]).tolist(),
+        replace=False,
+        size=number_samples,
+    )
+    chosen_indeces = [
+        index
+    ] + chosen_indeces.tolist()  # make sure sample under consideration included
+    data = data.iloc[chosen_indeces]
+    data = data.sort_index()
+    return data
+
+
+def build_constraints(data, i, keys_immutable):
+    """
+
+    Parameters
+    ----------
+    data: pd.DataFrame
+    i : int
+        Position of immutable key
+    keys_immutable: list[str]
+        Immutable feature
+
+    Returns
+    -------
+    np.ndarray, np.ndarray
+    """
+    epsilon = 0.5  # avoids division by 0
+    immutable_constraint_matrix = np.outer(
+        data[keys_immutable[i]].values + epsilon,
+        data[keys_immutable[i]].values + epsilon,
+    )
+    immutable_constraint_matrix1 = immutable_constraint_matrix / ((1 + epsilon) ** 2)
+    immutable_constraint_matrix1 = ((immutable_constraint_matrix1 == 1) * 1).astype(
+        float
+    )
+    immutable_constraint_matrix2 = immutable_constraint_matrix / (epsilon ** 2)
+    immutable_constraint_matrix2 = ((immutable_constraint_matrix2 == 1) * 1).astype(
+        float
+    )
+    return immutable_constraint_matrix1, immutable_constraint_matrix2
+
+
+def find_cf(
```

Comment:
Nitpick: No need to save characters here. More expressiveness/clarity almost always has more priotity over saving space. 
```suggestion
def find_counterfactuals(
```